Sources/NIO/CircularBuffer.swift

```diff
@@ -97,17 +178,36 @@ public struct CircularBuffer<E>: CustomStringConvertible, AppendableCollection {
         self.buffer = newBacking
     }
 
-    /// Remove the front element of the ring buffer.
+    /// Removes & returns the item at `position` from the buffer
     ///
-    /// *O(1)*
-    public mutating func removeFirst() -> E {
-        guard let value = self.buffer[self.headIdx] else {
-            preconditionFailure("CircularBuffer is empty")
+    /// - Parameter position: The index of the item to be removed from the buffer.
+    ///
+    /// *O(1)* if the position is `headIdx` or `tailIdx`.
+    /// otherwise
+    /// *O(n)* where *n* is the number of elements between `position` and `tailIdx`.
+    public mutating func remove(at position: Int) -> E {
+        var bufferIndex = self.bufferIndex(ofIndex: position)
+        let element = self.buffer[bufferIndex]!
+
+        switch bufferIndex {
+        case self.headIdx:
+            self.headIdx = (self.headIdx + 1) & self.mask
+            self.buffer[bufferIndex] = nil
+        case self.tailIdx - 1:
```

Comment:
This looks wrong to me: can you test this function including this case?

Reply:
Test can be found here https://github.com/esttorhe/swift-nio/blob/9133c12f9a7122741c0d8ae750e6a93d5bd7be83/Tests/NIOTests/CircularBufferTests.swift#L65